privacy/ccpa/policy.go

```diff
@@ -9,139 +9,190 @@ import (
+		regsResult.Ext = nil
+		return &regsResult, nil
 	}
 
-	var extMap map[string]interface{}
-	err := json.Unmarshal(req.Regs.Ext, &extMap)
+	// Marshal ext if there are still other fields
+	var regsResult openrtb.Regs
+	ext, err := json.Marshal(extMap)
 	if err == nil {
-		delete(extMap, "us_privacy")
-		if len(extMap) == 0 {
-			req.Regs.Ext = nil
-		} else {
-			ext, err := json.Marshal(extMap)
-			if err == nil {
-				req.Regs.Ext = ext
-			}
-			return err
-		}
+		regsResult = *regs
+		regsResult.Ext = ext
 	}
-
-	return err
+	return &regsResult, err
 }
 
-// Validate returns an error if the CCPA policy does not adhere to the IAB spec.
-func (p Policy) Validate() error {
-	if err := ValidateConsent(p.Value); err != nil {
-		return fmt.Errorf("request.regs.ext.us_privacy %s", err.Error())
+func buildRegsWrite(consent string, regs *openrtb.Regs) (*openrtb.Regs, error) {
+	if regs == nil {
+		return marshalRegsExt(openrtb.Regs{}, openrtb_ext.ExtRegs{USPrivacy: consent})
 	}
 
-	return nil
+	if regs.Ext == nil {
+		return marshalRegsExt(*regs, openrtb_ext.ExtRegs{USPrivacy: consent})
+	}
+
+	var extMap map[string]interface{}
+	if err := json.Unmarshal(regs.Ext, &extMap); err != nil {
+		return nil, err
+	}
+
+	extMap["us_privacy"] = consent
+	return marshalRegsExt(*regs, extMap)
 }
 
-// ValidateConsent returns an error if the CCPA consent string does not adhere to the IAB spec.
-func ValidateConsent(consent string) error {
-	if consent == "" {
-		return nil
+func marshalRegsExt(regs openrtb.Regs, ext interface{}) (*openrtb.Regs, error) {
+	extJSON, err := json.Marshal(ext)
+	if err == nil {
+		regs.Ext = extJSON
 	}
+	return &regs, err
+}
 
-	if len(consent) != 4 {
-		return errors.New("must contain 4 characters")
+func buildExt(noSaleBidders []string, ext json.RawMessage) (json.RawMessage, error) {
+	if len(noSaleBidders) == 0 {
+		return buildExtClear(ext)
 	}
+	return buildExtWrite(noSaleBidders, ext)
+}
 
-	if consent[0] != '1' {
-		return errors.New("must specify version 1")
+func buildExtClear(ext json.RawMessage) (json.RawMessage, error) {
+	if len(ext) == 0 {
+		return ext, nil
 	}
 
-	var c byte
+	var extMap map[string]interface{}
+	if err := json.Unmarshal(ext, &extMap); err != nil {
+		return nil, err
+	}
 
-	c = consent[1]
-	if c != 'N' && c != 'Y' && c != '-' {
-		return errors.New("must specify 'N', 'Y', or '-' for the explicit notice")
+	prebidExt, exists := extMap["prebid"]
+	if !exists {
+		return ext, nil
 	}
 
-	c = consent[2]
-	if c != 'N' && c != 'Y' && c != '-' {
-		return errors.New("must specify 'N', 'Y', or '-' for the opt-out sale")
+	// Verify prebid is an object
+	prebidExtMap, ok := prebidExt.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("request.ext.prebid is not a json object")
 	}
 
-	c = consent[3]
-	if c != 'N' && c != 'Y' && c != '-' {
-		return errors.New("must specify 'N', 'Y', or '-' for the limited service provider agreement")
+	// Remove no sale member
+	delete(prebidExtMap, "nosale")
```

Comment:
How is this modified `prebidExtMap` being reassigned to `extMap`? I see the test cases pass, particularly the one where `nosale` is removed but other fields in `prebid` remain...